torch/distributed/distributed_c10d.py

```diff
@@ -519,14 +531,35 @@ class GroupMember(metaclass=_WorldMeta):
 STORE_BASED_BARRIER_PREFIX = "store_based_barrier_key"
 
 
-def _get_pg_device(group: ProcessGroup):
+def _get_object_coll_device(group: Optional[ProcessGroup] = None):
+    group = group or _get_default_group()
+    """
+    ``group._device_types`` is a property pybind that returns the devices
+    ("cpu", "cuda", etc) supported by ``group``. Can be multiple if the
+    ``group`` supports multiple devices.
+    """
+    devices = group._device_types
+
+    if len(devices) == 1:
+        # User fixed exactly one backend in `init_process_group`
+        return devices[0]
+    elif len(devices) == 0:
+        # No backend has been registered with this PG (maybe because no
+        # collective has been run?) We pick cpu as the default and hopefully
+        # this would lazily init Gloo or other available cpu backend.
+        return torch.device("cpu")
+    elif torch.device("cpu") in devices:
+        # There are multiple backends in this PG and cpu is among them.
+        # cpu is preferred as the object is in cpu memory. No need for device
+        # copy.
+        return torch.device("cpu")
+    else:
+        # No cpu in the backend list. Randomly pick the first backend
+        return devices[0]
     """
-    Returns the device to use with ``group``.
-    This is cuda for NCCL and CPU for everything else
+    TODO: see if we can cache this object_coll_device so that we only need to
```

Comment:
could we just add a decorator? https://docs.python.org/3/library/functools.html#functools.lru_cache